### PR TITLE
Attaching Exception records with pending payments to the case

### PR DESCRIPTION
**Pull request by aliveni**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1025

### Change description ###
Allow attaching exception record with pending payments to the case for the configured journey classifications.
Functional tests will be added in the next pr. 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by rkondratowicz** [comment]
From acceptance criteria on story:
>Functional tests have to be merged not later than the PR that changes the behaviour of the callback endpoint.

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java [hunk 1]
```diff
@@ -198,12 +199,40 @@ private CallbackValidations() {
...
+    static Validation<String, Void> validatePayments(
+        CaseDetails theCase,
+        Classification classification,
+        ServiceConfigItem config
+    ) {
+        Optional<String> awaitingPaymentsOptional = getAwaitingPaymentDcnProcessing(theCase);
+
+        if (!awaitingPaymentsOptional.isPresent()
+            || (awaitingPaymentsOptional.get().equals("No") // no payments or no payments pending for completion
+                || (awaitingPaymentsOptional.get().equals("Yes") // payments processing pending
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java]: by any chance, do we have string constants for ccd yes/no?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java [hunk 2]
```diff
@@ -198,12 +199,40 @@ private CallbackValidations() {
...
+                    && config.getAllowAttachingToCaseBeforePaymentsAreProcessedForClassifications()
+                        .contains(classification)))
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java]: really hard to follow, maybe it would be clearer with multiple if-else or nested ifs

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/TestCaseBuilder.java [hunk 3]
```diff
@@ -46,6 +46,16 @@ static CaseDetails caseWithSearchCaseReference(Object searchCaseReferenceType, O
         return createCaseWith(b -> b.data(caseData));
     }
 
+    static CaseDetails caseWithAwaitingPaymentsAndClassification(
+        Object awaitingPaymentsProcessing,
+        Object classification
+    ) {
+        Map<String, Object> data = new HashMap<>();
+        data.put("awaitingPaymentDCNProcessing", awaitingPaymentsProcessing);
+        data.put("journeyClassification", classification);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/TestCaseBuilder.java]: are there constants for those somewhere?

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java [hunk 4]
```diff
@@ -198,12 +199,40 @@ private CallbackValidations() {
...
+        ) {
+            return valid(null);
+        } else {
+            return invalid(
+                "The 'attach to case' event is not supported for the Exception Record with pending payments"
+            );
+        }
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java]: that main if statement.. maybe we could reverse it and approach in functional way?

```java
return getAwaitingPaymentDcnProcessing(theCase)
    .filter(flag -> !flag.equals("No"))
    .filter(flag -> !(flag.equals("Yes") && config.contains(classification))
    .map(voidFlag -> invalid(...))
    .orElse(valid(null));
```

p.s. comments are great for clarity ^^ ;)

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java (hunk 2 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java]: how about this?
```
 if (awaitingPaymentsOptional.isPresent()
            && awaitingPaymentsOptional.get().equals(YES) // payments processing pending
            && !config.getAllowAttachingToCaseBeforePaymentsAreProcessedForClassifications() // check if config allows
            .contains(classification)
        ) {
            return invalid(
                "The 'attach to case' event is not supported for the Exception Record with pending payments"
            );
        } else {
            return valid(null);
        }
```

**Comment by aliveni** [comment]
> From acceptance criteria on story:
> 
> > Functional tests have to be merged not later than the PR that changes the behaviour of the callback endpoint.

I'm not merging this PR until the functional tests are ready

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java (hunk 2 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java]: much clearer!

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java (hunk 1 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java]: yes, updated

File: src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/TestCaseBuilder.java (hunk 3 above)
Comment (aliveni) [reply src/test/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/TestCaseBuilder.java]: yes, updated with constants

File: src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java (hunk 4 above)
Comment (aliveni) [reply src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java]: thanks, but the updated  IF statement is much simpler

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/AttachExceptionRecordToExistingCaseTest.java [hunk 5]
```diff
@@ -580,7 +584,99 @@ public void should_fail_when_classification_is_supplementary_evidence_with_ocr_d
...
+            .then()
+            .statusCode(200)
+            .body(
+                RESPONSE_FIELD_ERRORS,
+                hasItem("Cannot attach this exception record to a case because it has pending payments")
+            );
+    }
+
+    @Test
+    public void should_succeed_when_classification_is_supplementary_evidence_with_processed_payments() {
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/AttachExceptionRecordToExistingCaseTest.java]: this test class now has 1111 lines of code. Maybe it would be worth moving the 'pending payments' tests to a new class?

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/AttachExceptionRecordToExistingCaseTest.java [hunk 6]
```diff
@@ -580,7 +584,99 @@ public void should_fail_when_classification_is_supplementary_evidence_with_ocr_d
...
+            .then()
+            .statusCode(200)
+            .body(
+                RESPONSE_FIELD_ERRORS,
+                hasItem("Cannot attach this exception record to a case because it has pending payments")
+            );
+    }
+
+    @Test
+    public void should_not_return_payments_error_when_supplementary_evidence_with_ocr_have_pending_payments() {
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/AttachExceptionRecordToExistingCaseTest.java]: it's not clear that this applies only to 'bulkscan' jurisdiction/service and that the rules on which classifications are allowed and which are not.
Is there are way to set up this in test in the 'given' stage?

Or at least providing the config on the test class? (extracting those tests to a separate class would come in handy here)

Comment (aliveni) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/AttachExceptionRecordToExistingCaseTest.java]: I'll move these tests to a separate class and rename the tests

Comment (aliveni) [reply src/integrationTest/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/AttachExceptionRecordToExistingCaseTest.java]: As discussed yesterday, I'll refactor the tests in a separate PR. For now, I renamed the tests and added a comment for classification.
